src/appleseed/renderer/kernel/lighting/volumelightingintegrator.cpp

```diff
@@ -0,0 +1,601 @@
+#include "renderer/kernel/lighting/tracer.h"
+#include "renderer/kernel/shading/shadingcomponents.h"
+#include "renderer/kernel/shading/shadingcontext.h"
+#include "renderer/kernel/shading/shadingpoint.h"
+#include "renderer/modeling/volume/volume.h"
+#include "renderer/modeling/edf/edf.h"
+#include "renderer/modeling/light/light.h"
+#include "renderer/modeling/material/material.h"
+#include "renderer/modeling/scene/visibilityflags.h"
+
+// appleseed.foundation headers.
+#include "foundation/math/rr.h"
+#include "foundation/math/sampling/mappings.h"
+#include "foundation/math/scalar.h"
+
+// Standard headers.
+#include <cassert>
+#include <cmath>
+
+using namespace foundation;
+
+namespace renderer
+{
+
+namespace
+{
+    struct EquiangularSampler
+    {
+      public:
+        EquiangularSampler(
+            const LightSample&          light_sample,
+            const ShadingRay&           volume_ray,
+            const ShadingContext&       shading_context,
+            SamplingContext&            sampling_context)
+          : m_shading_context(shading_context)
+          , m_sampling_context(sampling_context)
+        {
+            Vector3d emission_position;
+            if (light_sample.m_light != nullptr)
+            {
+                m_sampling_context.split_in_place(2, 1);
+                Vector3d emission_direction;  // not used
+                Spectrum light_value(Spectrum::Illuminance);
+                float probability;
+                light_sample.m_light->sample(
+                    m_shading_context,
+                    light_sample.m_light_transform,
+                    m_sampling_context.next2<Vector2d>(),
+                    emission_position,
+                    emission_direction,
+                    light_value,
+                    probability);
+            }
+            else
+                emission_position = light_sample.m_point;
+
+            const Vector3d light_dir = emission_position - volume_ray.m_org;
+
+            m_origin_to_center = dot(light_dir, volume_ray.m_dir);
+
+            m_center_to_light =
+                std::sqrt(square_norm(light_dir) - square(m_origin_to_center));
+
+            m_near_angle = std::atan2(-m_origin_to_center, m_center_to_light);
+            m_far_angle =
+                (volume_ray.m_tmax == std::numeric_limits<double>::max()) ? HalfPi<double>() :
+                std::atan2(volume_ray.get_length() - m_origin_to_center, m_center_to_light);
+        }
+
+        float sample() const
+        {
+            SamplingContext child_sampling_context = m_sampling_context.split(1, 1);
+            const double s = child_sampling_context.next2<double>();
+
+            return static_cast<float>(m_origin_to_center +
+                sample_equiangular_distribution(s, m_near_angle, m_far_angle, m_center_to_light));
+        }
+
+        float evaluate(const float distance_sample) const
+        {
+            return static_cast<float>(equiangular_distribution_pdf(
```

Comment:
My advice here: use `float` everywhere. That means changing the type of the four member variables to `float` and adding static casts where you compute them, in the constructor.